cylc/flow/network/resolvers.py

```diff
@@ -194,15 +194,18 @@ def node_ids_filter(tokens, state, items) -> bool:
             )
             # match cycle/task/job state
             and (
-                not get_state_from_selectors(item)
+                not (
+                    state is not None
+                    and get_state_from_selectors(item)
+                )
                 or get_state_from_selectors(item) == state
             )
         )
         for item in uniq(items)
     )
 
 
-def node_filter(node, node_type, args):
+def node_filter(node, node_type, args, state):
```

Comment:
I'm not too clear on which bits belong to the node and which bits belong to the query?

Is the "state" the node state or the query state?

Reply:
`args` will have the query state to filter against
`state` is the node state or data-store node state (if state isn't set in the delta, i.e. with outputs).

This ensures we always have a state to filter against (i.e. if we ever wanted to filter deltas by state, then deltas without a state field set can still be filtered by their data-store state)

Reply:
Cheers, so would this be correct?:

```python
def node_filter(node, node_type, args, state):
    """Filter nodes based on attribute arguments

    Args:
        node: The node to filter (from the data or delta store).
        node_type: The type of the node being filtered (or is it the type to filter by?)
        args: The query arguments.
        state: The state of the node that is being filtered.
    """
```

Reply:
Yes, that's right.